qiskit/transpiler/passes/synthesis/plugin.py

```diff
@@ -0,0 +1,283 @@
+
+If for some reason the available inputs to the
+:meth:`~qiskit.transpiler.passes.synthesis.plugin.UnitarySynthesisPlugin.run`
+method are insufficient please open an issue and we can discuss expanding the
+plugin interface with new opt-in inputs that can be added in a backwards
+compatible manner for future releases. Do note though that this plugin interface
+is considered stable and guaranteed to not change in a breaking manner. If
+changes are needed (for example to expand the available optional input options)
+it will be done in a way that will **not** require changes from existing
+plugins.
+
+The second step is to expose the
+:class:`~qiskit.transpiler.passes.synthesis.plugin.UnitarySynthesisPlugin` as
+a setuptools entry point in the package metadata. This is done by simply adding
+an ``entry_points`` entry to the ``setuptools.setup`` call in the ``setup.py``
+for the plugin package with the necessary entry points under the
+``qiskit.unitary_synthesis`` namespace. For example::
+
+    entry_points={
+        'qiskit.unitary_synthesis': [
+            'special = qiskit_plugin_pkg.module.plugin:SpecialUnitarySynthesis',
+        ]
+    },
+
+(note that the entry point ``name = path`` is a single string not a Python
+expression). There isn't a limit to the number of plugins a single package can
+include as long as each plugin has a unique name. So a single package can
+expose multiple plugins if necessary.
+
+Using Plugins
+=============
+
+To use a plugin all you need to do is install the package that includes a
+synthesis plugin. Then Qiskit will automatically discover the installed
+plugins and expose them as valid options for the appropriate
+:func:`~qiskit.compiler.transpiler` kwargs and pass constructors. If there are
+any installed plugins which can't be loaded/imported this will be logged to
+Python logging.
+
+To get the installed list of installed unitary synthesis plugins you can use the
+:func:`qiskit.transpiler.passes.synthesis.plugin.unitary_synthesis_plugin_names`
+function.
+
+Plugin API
+==========
+
+Unitary Synthesis Plugins
+-------------------------
+
+.. autosummary::
+   :toctree: ../stubs/
+
+   UnitarySynthesisPlugin
+   UnitarySynthesisPluginManager
+   unitary_synthesis_plugin_names
+
+"""
+
+import abc
+
+import stevedore
+
+
+class UnitarySynthesisPlugin(abc.ABC):
+    """Abstract plugin Synthesis plugin class
+
+    This abstract class defines the interface for unitary synthesis plugins.
+    """
+
+    @property
+    @abc.abstractmethod
+    def supports_basis_gates(self):
+        """Return whether the plugin supports taking ``basis_gates``"""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_coupling_map(self):
+        """Return whether the plugin supports taking ``coupling_map``"""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_approximation_degree(self):
+        """Return whether the plugin supports taking ``approximation_degree``"""
+        pass
```

Comment:
Do we want to say anything more about the format that these arguments will be provided in?  Later abstract properties say "refer to the UnitarySynthesisPass" documentation - if we want to say that for all properties, we can probably just stick it in the main class docstring.